### PR TITLE
[XPU] add speculate_get_logits

### PR DESCRIPTION
**Pull request by RuohengMa**
## Motivation

add speculate_get_logits

## Modifications

add speculate_get_logits

## Usage or Command

No

## Accuracy Tests

No

## Checklist

- [ ] Add at least a tag in the PR title.
  - Tag list: [`[FDConfig]`,`[APIServer]`,`[Engine]`, `[Scheduler]`, `[PD Disaggregation]`, `[Executor]`, `[Graph Optimization]`, `[Speculative Decoding]`, `[RL]`, `[Models]`, `[Quantization]`, `[Loader]`, `[OP]`, `[KVCache]`, `[DataProcessor]`, `[BugFix]`, `[Docs]`, `[CI]`, `[Optimization]`, `[Feature]`, `[Benchmark]`, `[Others]`, `[XPU]`, `[HPU]`, `[GCU]`, `[DCU]`, `[Iluvatar]`, `[Metax]`]
  - You can add new tags based on the PR content, but the semantics must be clear.
- [ ] Format your code, run `pre-commit` before commit.
- [ ] Add unit tests. Please write the reason in this PR if no unit tests.
- [ ] Provide accuracy results.
- [ ] If the current PR is submitting to the `release` branch, make sure the PR has been submitted to the `develop` branch, then cherry-pick it to the `release` branch with the `[Cherry-Pick]` PR tag.

### CODE REVIEW

File: custom_ops/xpu_ops/src/ops/mtp/speculate_get_logits.cc [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+                        const paddle::Tensor& logits,
+                        const paddle::Tensor& first_token_logits,
+                        const paddle::Tensor& seq_lens_this_time,
+                        const paddle::Tensor& seq_lens_encoder) {
+  phi::XPUPlace place(phi::backends::xpu::GetXPUCurrentDeviceId());
+  auto dev_ctx = paddle::experimental::DeviceContextPool::Instance().Get(place);
+  baidu::xpu::api::Context* ctx =
+      static_cast<const phi::XPUContext*>(dev_ctx)->x_context();
+  if (draft_logits.is_cpu()) {
+    ctx = new baidu::xpu::api::Context(baidu::xpu::api::kCPU);
```
Comment (hong19860320) [inline custom_ops/xpu_ops/src/ops/mtp/speculate_get_logits.cc:41]: 这个 context 什么时候被销毁掉呢？是否会造成内存泄露？

**Review by hong19860320 — CHANGES_REQUESTED** [review]

**Review by cmcamdy — DISMISSED** [review]

File: custom_ops/xpu_ops/src/ops/mtp/speculate_get_logits.cc (hunk 1 above)
Comment (cmcamdy) [reply custom_ops/xpu_ops/src/ops/mtp/speculate_get_logits.cc:41]: CPU一般是用于单测验证，除了这里其他的算子可能也没对这个cpu的ctx做释放，后续可能需要统一排查一下

**Review by hong19860320 — DISMISSED** [review]
LGTM

**Review by cmcamdy — DISMISSED** [review]
LGTM

File: custom_ops/xpu_ops/src/plugin/src/wrapper/mtp_wrapper/speculate_get_logits.cpp [hunk 2]
```diff
@@ -0,0 +1,176 @@
...
+                      cu_next_token_offset,
+                      cu_batch_token_offset,
+                      logits);
+  WRAPPER_DUMP_PARAM5(ctx,
+                      first_token_logits,
+                      seq_lens_this_time,
+                      seq_lens_encoder,
+                      real_bsz,
+                      vocab_size);
+  WRAPPER_DUMP(ctx);
```
Comment (mayang002) [inline custom_ops/xpu_ops/src/plugin/src/wrapper/mtp_wrapper/speculate_get_logits.cpp:141]: 对输入输出的 xpu 指针加下 chekc 检查？

File: custom_ops/xpu_ops/src/plugin/src/kernel/kunlun3cpp/mtp_kernel/speculate_get_logits.xpu [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+          sm_seq_lens_encoder[bid] > 0 ? 1 : sm_seq_lens_this_time[bid];
+    }
+    mfence_sm();
+    if (clus_id == 0) {
+      SM2GM_ASYNC(sm_batch_token_num, batch_token_num, real_bsz * sizeof(int));
+      SM2GM_ASYNC(sm_cu_batch_token_offset,
+                  cu_batch_token_offset,
+                  real_bsz * sizeof(int));
+    }
+  }
```
Comment (mayang002) [inline custom_ops/xpu_ops/src/plugin/src/kernel/kunlun3cpp/mtp_kernel/speculate_get_logits.xpu:47]: 这个部分的代码逻辑，21行-40行是每个 cluster 都会执行，41-46行只有 cluster0 会执行，是不是等价于21行-46行的代码实际上只有 clus_id == 0 执行的有用？

**Review by mayang002 — CHANGES_REQUESTED** [review]

File: custom_ops/xpu_ops/src/plugin/src/kernel/kunlun3cpp/mtp_kernel/speculate_get_logits.xpu (hunk 3 above)
Comment (RuohengMa) [reply custom_ops/xpu_ops/src/plugin/src/kernel/kunlun3cpp/mtp_kernel/speculate_get_logits.xpu:47]: 这里的prefix sum所有参与计算的cluster都需要使用，所以21-40所有< real_bsz的cluster都要计算一份；但是写回gm的话，只要一个cluster写就行，所以41-46只要cluster0

File: custom_ops/xpu_ops/src/plugin/src/wrapper/mtp_wrapper/speculate_get_logits.cpp (hunk 2 above)
Comment (RuohengMa) [reply custom_ops/xpu_ops/src/plugin/src/wrapper/mtp_wrapper/speculate_get_logits.cpp:141]: 好的

File: custom_ops/xpu_ops/src/plugin/src/kernel/kunlun3cpp/mtp_kernel/speculate_get_logits.xpu (hunk 3 above)
Comment (mayang002) [reply custom_ops/xpu_ops/src/plugin/src/kernel/kunlun3cpp/mtp_kernel/speculate_get_logits.xpu:47]: > 这里的prefix sum所有参与计算的cluster都需要使用，所以21-40所有< real_bsz的cluster都要计算一份；但是写回gm的话，只要一个cluster写就行，所以41-46只要cluster0

好的，明白了

**Review by mayang002 — APPROVED** [review]
